scapy/contrib/stamp.py

```diff
@@ -0,0 +1,311 @@
+# SPDX-License-Identifier: GPL-2.0-only
+# This file is part of Scapy
+# See https://scapy.net/ for more information
+# Copyright (C) Carmine Scarpitta <carmine.scarpitta@uniroma2.it>
+
+# scapy.contrib.description = Simple Two-Way Active Measurement Protocol (STAMP)
+# scapy.contrib.status = loads
+
+"""
+STAMP (Simple Two-Way Active Measurement Protocol) - RFC 8762.
+
+References:
+    * `Simple Two-Way Active Measurement Protocol [RFC 8762]
+      <https://www.rfc-editor.org/rfc/rfc8762.html>`_
+    * `Simple Two-Way Active Measurement Protocol Optional Extensions [RFC 8972]
+      <https://www.rfc-editor.org/rfc/rfc8972.html>`_
+
+Remember to use::
+
+    bind_layers(UDP, STAMPSessionReflectorTestUnauthenticated, dport=XXX)
+
+to register the port you are using to receive STAMP Session-Reflector Test packets.
+"""
+
+from scapy import config
+from scapy.base_classes import Packet_metaclass
+from scapy.layers.inet import UDP
+from scapy.layers.ntp import TimeStampField
+from scapy.packet import Packet, bind_layers
+from scapy.fields import (
+    BitEnumField,
+    BitField,
+    ByteEnumField,
+    ByteField,
+    FlagsField,
+    IntField,
+    MultipleTypeField,
+    NBytesField,
+    PacketField,
+    PacketListField,
+    ShortField,
+    StrLenField,
+    UTCTimeField
+)
+
+
+_sync_types = {
+    0: 'No External Synchronization for the Time Source',
+    1: 'Clock Synchronized to UTC using an External Source'
+}
+
+_timestamp_types = {
+    0: 'NTP 64-bit Timestamp Format',
+    1: 'PTPv2 Truncated Timestamp Format'
+}
+
+_stamp_tlvs = {
+
```

Comment:
Is it expected that this is empty?

Reply:
Yes, in this PR it is supposed to be empty because I decided not to support any TLV.

According to RFC 8972 (Simple Two-Way Active Measurement Protocol Optional Extensions) TLVs are optional:

> The Type-Length-Value (TLV) encoding scheme provides a flexible extension mechanism for
> optional informational elements. TLV is an optional field in the STAMP test packet.

RFC 8972 defines several TLVs that can be **optionally** inserted into the packet.

I decided not to add any TLV in this PR to keep the PR small and simplify the review process.
However, I put some code that allows us to add TLVs to the STAMP layer in the future easily.

In particular, I added an optional field `tlv_objects` to the STAMP Test packets and implemented a Generic TLV packet (`STAMPTestTLV`).

To add a TLV in the future, you just need to:
* add a new class that extends the Generic TLV packet;
* add a new entry to the `_stamp_tlvs` dictionary, which associates the TLV type value with a description.